File: app/src/main/java/com/kongjak/ggcj/Activity/MainActivity.java
package com.kongjak.ggcj.Activity;

import android.app.ProgressDialog;
import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.content.SharedPreferences;
import android.net.Uri;
import android.os.AsyncTask;
import android.os.Bundle;
import android.support.customtabs.CustomTabsIntent;
import android.support.design.widget.NavigationView;
import android.support.v4.app.ActivityCompat;
import android.support.v4.view.GravityCompat;
import android.support.v4.widget.DrawerLayout;
import android.support.v7.app.ActionBarDrawerToggle;
import android.support.v7.app.AlertDialog;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.Toolbar;
import android.util.Log;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.widget.TextView;

import com.kongjak.ggcj.R;
import com.kongjak.ggcj.Tools.CheckDigit;
import com.kongjak.ggcj.Tools.ParseMeal;
import com.kongjak.ggcj.Tools.ParseSchedule;

import java.text.SimpleDateFormat;
import java.util.Calendar;
import java.util.Date;
import java.util.Locale;
import java.util.TimeZone;

public class MainActivity extends AppCompatActivity
        implements NavigationView.OnNavigationItemSelectedListener {

    private String getYear, getMonth, getDay, getDate;
    private String next_schedule_day = "", next_schedule = "";

    private AlertDialog alertDialog;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        checkFirstRun();

        Toolbar toolbar = findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);

        DrawerLayout drawer = findViewById(R.id.drawer_layout);
        ActionBarDrawerToggle toggle = new ActionBarDrawerToggle(
                this, drawer, toolbar, R.string.navigation_drawer_open, R.string.navigation_drawer_close);
        drawer.addDrawerListener(toggle);
        toggle.syncState();

        NavigationView navigationView = findViewById(R.id.nav_view);
        navigationView.setNavigationItemSelectedListener(this);

        navigationView.setCheckedItem(R.id.nav_home);

        IntentFilter filter = new IntentFilter();
        filter.addAction("MealParseEnd");
        filter.addAction("ScheduleParseEnd");
        registerReceiver(mBroadcastReceiver, filter);

        long now = System.currentTimeMillis();
        Date date = new Date(now);

        SimpleDateFormat year = new SimpleDateFormat("yyyy", Locale.KOREA);
        getYear = year.format(date);

        SimpleDateFormat month = new SimpleDateFormat("MM", Locale.KOREA);
        getMonth = CheckDigit.check(month.format(date));

        SimpleDateFormat day = new SimpleDateFormat("dd", Locale.KOREA);
        getDay = CheckDigit.check(day.format(date));

        setView();
    }

    private void checkFirstRun() {
        SharedPreferences sp = getSharedPreferences("AppIntro", MODE_PRIVATE);
        if (!sp.getBoolean("first", false)) {
            SharedPreferences.Editor editor = sp.edit();
            editor.putBoolean("first", true);
            editor.apply();
            Intent intent = new Intent(this, IntroActivity.class); // Call the AppIntro java class
            startActivity(intent);
        }
    }

    private void setView() {
        getDate = getYear + "-" + getMonth + "-" + getDay;
        checkMeal();
        checkSchedule();
    }

    private void checkMeal() {
        SharedPreferences lunch_sp = getSharedPreferences("lunch", MODE_PRIVATE);
        String lunch_str = lunch_sp.getString(getDate, "");

        if (!lunch_str.isEmpty()) {
            setMealView();
        } else {
            EmptyDialog(getString(R.string.meal));
        }
    }

    private void setMealView() {
        SharedPreferences lunch_sp = getSharedPreferences("lunch", MODE_PRIVATE);
        String lunch_str = lunch_sp.getString(getDate, "");

        SharedPreferences dinner_sp = getSharedPreferences("dinner", MODE_PRIVATE);
        String dinner_str = dinner_sp.getString(getDate, "");

        TextView lunch_v = findViewById(R.id.item_lunch);
        lunch_v.setVisibility(View.VISIBLE);

        TextView dinner_v = findViewById(R.id.item_dinner);
        dinner_v.setVisibility(View.VISIBLE);

        lunch_v.setText(lunch_str);
        dinner_v.setText(dinner_str);
    }

    private void checkSchedule() {
        SharedPreferences schedule_sp = getSharedPreferences("schedule", MODE_PRIVATE);
        String schedule_str = schedule_sp.getString(getDate, "");

        TextView schedule_v = findViewById(R.id.item_schedule);
        schedule_v.setVisibility(View.VISIBLE);

        if (!schedule_str.isEmpty()) {
            setScheduleView();
        } else {
            EmptyDialog(getString(R.string.schedule));
        }
    }

    private void setScheduleView() {
        SharedPreferences schedule_sp = getSharedPreferences("schedule", MODE_PRIVATE);
        String schedule_str = schedule_sp.getString(getDate, "");

        TextView schedule_v = findViewById(R.id.item_schedule);
        schedule_v.setVisibility(View.VISIBLE);

        if (schedule_str.equals(getString(R.string.no_schedule))) {
            NextSchedule asyncTask = new NextSchedule();
            asyncTask.execute();
        } else {
            schedule_v.setText(schedule_str);
        }
    }

    @Override
    public void onBackPressed() {
        DrawerLayout drawer = findViewById(R.id.drawer_layout);
        if (drawer.isDrawerOpen(GravityCompat.START)) {
            drawer.closeDrawer(GravityCompat.START);
        } else {
            ActivityCompat.finishAffinity(this);
            super.onBackPressed();
        }
    }

    @SuppressWarnings("StatementWithEmptyBody")
    @Override
    public boolean onNavigationItemSelected(MenuItem item) {
        // Handle navigation view item clicks here.
        int id = item.getItemId();

        if (id == R.id.nav_notice) {
            Intent intent = new Intent(getBaseContext(), NoticeActivity.class);
            intent.putExtra("url", getString(R.string.notice_url));
            intent.putExtra("type", 0);
            startActivity(intent);
        } else if (id == R.id.nav_notice_parents) {
            Intent intent = new Intent(getBaseContext(), NoticeActivity.class);
            intent.putExtra("url", getString(R.string.notice_parents_url));
            intent.putExtra("type", 1);
            startActivity(intent);
        } else if (id == R.id.nav_date) {
            Intent intent = new Intent(getBaseContext(), DateActivity.class);
            startActivity(intent);
        } else if (id == R.id.nav_gallery) {
            Intent intent = new Intent(getBaseContext(), GalleryActivity.class);
            startActivity(intent);
        } else if (id == R.id.nav_timetable) {
            String url = "http://comci.kr/st";
            CustomTabsIntent.Builder builder = new CustomTabsIntent.Builder();
            CustomTabsIntent customTabsIntent = builder.build();
            customTabsIntent.launchUrl(MainActivity.this, Uri.parse(url));
        } else if (id == R.id.nav_send) {
            Intent email = new Intent(Intent.ACTION_SENDTO);
            email.setData(Uri.parse("mailto:"));
            String[] address = {"ggcj@kongjak.com"};
            email.putExtra(Intent.EXTRA_EMAIL, address);
            String title = getResources().getString(R.string.nav_send_intent);
            Intent chooser = Intent.createChooser(email, title);
            if (email.resolveActivity(getPackageManager()) != null) {
                startActivity(chooser);
            }
        } else if (id == R.id.nav_info) {
            Intent intent = new Intent(getBaseContext(), InfoActivity.class);
            startActivity(intent);
        }

        DrawerLayout drawer = findViewById(R.id.drawer_layout);
        drawer.closeDrawer(GravityCompat.START);
        return true;
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        // Inflate the menu; this adds items to the action bar if it is present.
        getMenuInflater().inflate(R.menu.main, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        // Handle action bar item clicks here. The action bar will
        // automatically handle clicks on the Home/Up button, so long
        // as you specify a parent activity in AndroidManifest.xml.
        int id = item.getItemId();

        //noinspection SimplifiableIfStatement
        if (id == R.id.action_web) {
            String url = "http://ggcj.hs.kr";
            CustomTabsIntent.Builder builder = new CustomTabsIntent.Builder();
            CustomTabsIntent customTabsIntent = builder.build();
            customTabsIntent.launchUrl(MainActivity.this, Uri.parse(url));
        }

        return super.onOptionsItemSelected(item);
    }

    private void EmptyDialog(String dataType) {
        AlertDialog.Builder builder = new AlertDialog.Builder(this);
        builder.setTitle(String.format(getString(R.string.data_empty_dialog), dataType));
        builder.setPositiveButton(getString(R.string.ok), (dialog, id) -> {
            if (dataType.equals(getString(R.string.meal))) {
                getDatas(0);
            } else if (dataType.equals(getString(R.string.schedule))) {
                getDatas(1);
            }
        });

        builder.setNegativeButton(getString(R.string.cancel), (dialog, id) -> dialog.dismiss());
        alertDialog = builder.create();
        alertDialog.show();
    }

    private void getDatas(int type) {
        if (type == 0) {
            ParseMeal.WeekMealTask lunchTask = new ParseMeal.WeekMealTask(MainActivity.this);
            lunchTask.execute("2", getYear, getMonth, getDay);

            ParseMeal.WeekMealTask dinnerTask = new ParseMeal.WeekMealTask(MainActivity.this);
            dinnerTask.execute("3", getYear, getMonth, getDay);
        } else if (type == 1) {
            ParseSchedule.ScheduleTask asyncTask = new ParseSchedule.ScheduleTask(MainActivity.this);
            asyncTask.execute(getYear, getMonth);
        }
    }

    private class NextSchedule extends AsyncTask<String, String, Void> {
        ProgressDialog asyncDialog = new ProgressDialog(
                MainActivity.this);

        @Override
        protected void onPostExecute(Void result) {
            //doInBackground 작업이 끝나고 난뒤의 작업
            TextView schedule_v = findViewById(R.id.item_schedule);
            if (next_schedule.isEmpty()) {
                schedule_v.setText(getString(R.string.no_next_schedule));
            } else {
                schedule_v.setText(String.format(getString(R.string.next_schedule), next_schedule, next_schedule_day));
            }
            asyncDialog.dismiss();
            super.onPostExecute(result);
        }

        @Override
        protected void onPreExecute() {
            String loading = getString(R.string.loading);
            asyncDialog.setProgressStyle(ProgressDialog.STYLE_SPINNER);
            asyncDialog.setMessage(loading);

            // show dialog
            asyncDialog.show();
            super.onPreExecute();
        }

        @Override
        protected Void doInBackground(String... params) {
            //백그라운드 작업이 진행되는 곳.
            int today = Calendar.getInstance(TimeZone.getDefault()).get(Calendar.DAY_OF_MONTH);
            int last_day = Calendar.getInstance().getActualMaximum(Calendar.DAY_OF_MONTH);
            try {
                for (int i = today + 1; i <= last_day; i++) {
                    String getFullDate = getYear + "-" + getMonth + "-" + CheckDigit.check(String.valueOf(i));
                    String getDate = getMonth + "/" + CheckDigit.check(String.valueOf(i));

                    SharedPreferences schedule_sp = getSharedPreferences("schedule", MODE_PRIVATE);
                    String schedule_str = schedule_sp.getString(getFullDate, "");

                    if (!schedule_str.equals(getString(R.string.no_schedule))) {
                        publishProgress(getDate, schedule_str); // Send it!
                        break;
                    }

                }
            } catch (ClassCastException e) {
                e.printStackTrace();
            }
            return null;
        }

        @Override
        protected void onProgressUpdate(String... params) { // Receive from doInBackground
            next_schedule_day = params[0];
            next_schedule = params[1];
        }
    }

    private BroadcastReceiver mBroadcastReceiver = new BroadcastReceiver() {
        @Override
        public void onReceive(Context context, Intent intent) {
            if (intent.getAction().equals("MealParseEnd")) {
                setMealView();
            } else if (intent.getAction().equals("ScheduleParseEnd")) {
                setScheduleView();
            }
        }
    };

    @Override
    protected void onPause() {
        super.onPause();
    }

    @Override
    protected void onResume() {
        super.onResume();
        NavigationView navigationView = findViewById(R.id.nav_view);
        navigationView.setCheckedItem(R.id.nav_home);
        Log.d("GGCJ", "Resume");
        setMealView();
        setScheduleView();
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        unregisterReceiver(mBroadcastReceiver);
    }
}

File: app/src/main/java/com/kongjak/ggcj/Activity/MainActivity.kt
package com.kongjak.ggcj.Activity

import android.content.*
import android.net.Uri
import android.os.Bundle
import android.util.Log
import android.view.Menu
import android.view.MenuItem
import android.view.View
import androidx.appcompat.app.ActionBarDrawerToggle
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.app.AppCompatActivity
import androidx.browser.customtabs.CustomTabsIntent
import androidx.core.app.ActivityCompat
import androidx.core.view.GravityCompat
import com.google.android.material.navigation.NavigationView
import com.kongjak.ggcj.R
import com.kongjak.ggcj.Tools.CheckDigit.check
import com.kongjak.ggcj.Tools.ParseMeal.WeekMealTask
import com.kongjak.ggcj.Tools.ParseSchedule.ScheduleTask
import kotlinx.android.synthetic.main.activity_main.*
import kotlinx.android.synthetic.main.app_bar_main.*
import kotlinx.android.synthetic.main.content_main.*
import java.text.SimpleDateFormat
import java.util.*

class MainActivity : AppCompatActivity(), NavigationView.OnNavigationItemSelectedListener {
    private var getYear: String? = null
    private var getMonth: String? = null
    private var getDay: String? = null
    private var getDate: String? = null
    private var alertDialog: AlertDialog? = null
    private val mBroadcastReceiver: BroadcastReceiver = object : BroadcastReceiver() {
        override fun onReceive(context: Context, intent: Intent) {
            loadingProgress.visibility = View.GONE
            if (intent.action == "MealParseEnd") {
                setMealView()
            } else if (intent.action == "ScheduleParseEnd") {
                setScheduleView()
            }
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        setSupportActionBar(toolbar)
        val toggle = ActionBarDrawerToggle(
            this,
            drawer_layout,
            toolbar,
            R.string.navigation_drawer_open,
            R.string.navigation_drawer_close
        )
        drawer_layout.addDrawerListener(toggle)
        toggle.syncState()
        nav_view.setNavigationItemSelectedListener(this)
        nav_view.setCheckedItem(R.id.nav_home)
        val filter = IntentFilter()
        filter.addAction("MealParseEnd")
        filter.addAction("ScheduleParseEnd")
        registerReceiver(mBroadcastReceiver, filter)
        val now = System.currentTimeMillis()
        val date = Date(now)
        val year = SimpleDateFormat("yyyy", Locale.KOREA)
        getYear = year.format(date)
        val month = SimpleDateFormat("MM", Locale.KOREA)
        getMonth = check(month.format(date))
        val day = SimpleDateFormat("dd", Locale.KOREA)
        getDay = check(day.format(date))
        checkFirstRun()
    }

    private fun setView() {
        getDate = "$getYear-$getMonth-$getDay"
        checkMeal()
        checkSchedule()
    }

    private fun checkFirstRun() {
        val sp = getSharedPreferences("AppIntro", Context.MODE_PRIVATE)
        if (!sp.getBoolean("first", false)) {
            val intent = Intent(this, IntroActivity::class.java) // Call the AppIntro java class
            startActivity(intent)
        }
    }


    private fun checkMeal() {
        val lunch_sp = getSharedPreferences("lunch", Context.MODE_PRIVATE)
        val lunch_str = lunch_sp.getString(getDate, "")
        if (lunch_str!!.isNotEmpty()) {
            setMealView()
        } else {
            EmptyDialog(getString(R.string.meal))
        }
    }

    private fun setMealView() {
        val lunch_sp = getSharedPreferences("lunch", Context.MODE_PRIVATE)
        val lunch_str = lunch_sp.getString(getDate, "")
        val dinner_sp = getSharedPreferences("dinner", Context.MODE_PRIVATE)
        val dinner_str = dinner_sp.getString(getDate, "")
        item_lunch.visibility = View.VISIBLE
        item_dinner.visibility = View.VISIBLE
        item_lunch.text = lunch_str
        item_dinner.text = dinner_str
    }

    private fun checkSchedule() {
        val schedule_sp = getSharedPreferences("schedule", Context.MODE_PRIVATE)
        val schedule_str = schedule_sp.getString(getDate, "")
        item_schedule.visibility = View.VISIBLE
        if (schedule_str!!.isNotEmpty()) {
            setScheduleView()
        } else {
            EmptyDialog(getString(R.string.schedule))
        }
    }

    private fun setScheduleView() {
        val schedule_sp = getSharedPreferences("schedule", Context.MODE_PRIVATE)
        val schedule_str = schedule_sp.getString(getDate, "")
        item_schedule.visibility = View.VISIBLE
        if (schedule_str == getString(R.string.no_schedule)) {
            nextSchedule
        } else {
            item_schedule.text = schedule_str
        }
    }

    override fun onBackPressed() {
        if (drawer_layout.isDrawerOpen(GravityCompat.START)) {
            drawer_layout.closeDrawer(GravityCompat.START)
        } else {
            ActivityCompat.finishAffinity(this)
            super.onBackPressed()
        }
    }

    override fun onNavigationItemSelected(item: MenuItem): Boolean {
        // Handle navigation view item clicks here.
        val id = item.itemId
        if (id == R.id.nav_notice) {
            val intent = Intent(baseContext, NoticeActivity::class.java)
            intent.putExtra("url", getString(R.string.notice_url))
            intent.putExtra("type", 0)
            startActivity(intent)
        } else if (id == R.id.nav_notice_parents) {
            val intent = Intent(baseContext, NoticeActivity::class.java)
            intent.putExtra("url", getString(R.string.notice_parents_url))
            intent.putExtra("type", 1)
            startActivity(intent)
        } else if (id == R.id.nav_date) {
            val intent = Intent(baseContext, DateActivity::class.java)
            startActivity(intent)
            /**
             * Comment out because only logged in user can access board.
             * So, let's block Gallery Menu until they grant access permission.
             *
            } else if (id == R.id.nav_gallery) {
            val intent = Intent(baseContext, GalleryActivity::class.java)
            startActivity(intent)
             *
             **/
        } else if (id == R.id.nav_timetable) {
            val url = resources.getString(R.string.menu_timetable_url)
            val builder = CustomTabsIntent.Builder()
            val customTabsIntent = builder.build()
            customTabsIntent.launchUrl(this, Uri.parse(url))
        } else if (id == R.id.nav_send) {
            val email = Intent(Intent.ACTION_SEND).apply {
                type = "message/rfc822"
                putExtra(Intent.EXTRA_EMAIL, arrayOf("ggcj@kongjak.com"))
            }
            val chooser = Intent.createChooser(email, resources.getString(R.string.nav_send_intent))
            if (intent.resolveActivity(packageManager) != null) {
                startActivity(chooser)
            }
        } else if (id == R.id.nav_info) {
            val intent = Intent(baseContext, InfoActivity::class.java)
            startActivity(intent)
        }
        drawer_layout.closeDrawer(GravityCompat.START)
        return true
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        // Inflate the menu; this adds items to the action bar if it is present.
        menuInflater.inflate(R.menu.main, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        // Handle action bar item clicks here. The action bar will
        // automatically handle clicks on the Home/Up button, so long
        // as you specify a parent activity in AndroidManifest.xml.
        val id = item.itemId
        if (id == R.id.action_web) {
            val url = "http://ggcj.hs.kr"
            val builder = CustomTabsIntent.Builder()
            val customTabsIntent = builder.build()
            customTabsIntent.launchUrl(this, Uri.parse(url))
        }
        return super.onOptionsItemSelected(item)
    }

    private fun EmptyDialog(dataType: String) {
        val builder = AlertDialog.Builder(this)
        builder.setTitle(String.format(getString(R.string.data_empty_dialog), dataType))
        builder.setPositiveButton(getString(R.string.ok)) { _, _ ->
            if (dataType == getString(R.string.meal)) {
                getDatas(0)
            } else if (dataType == getString(R.string.schedule)) {
                getDatas(1)
            }
        }
        builder.setNegativeButton(getString(R.string.cancel)) { dialog: DialogInterface, _ -> dialog.dismiss() }
        alertDialog = builder.create()
        alertDialog!!.show()
    }

    private fun getDatas(type: Int) {
        loadingProgress.visibility = View.VISIBLE
        if (type == 0) {
            val lunchTask = WeekMealTask(this)
            lunchTask.execute("2", getYear, getMonth)
            val dinnerTask = WeekMealTask(this)
            dinnerTask.execute("3", getYear, getMonth)
        } else if (type == 1) {
            val asyncTask = ScheduleTask(this)
            asyncTask.execute(getYear, getMonth)
        }
    }

    // show dialog
    private val nextSchedule: Unit
        get() {
            val today = Calendar.getInstance(TimeZone.getDefault())[Calendar.DAY_OF_MONTH]
            val last_day = Calendar.getInstance().getActualMaximum(Calendar.DAY_OF_MONTH)
            loadingProgress.visibility = View.VISIBLE
            Thread {
                var nextSchedule: String? = null
                var nextScheduleDay: String? = null
                for (i in today + 1..last_day) {
                    val getFullDate = getYear + "-" + getMonth + "-" + check(i.toString())
                    val getDate = getMonth + "/" + check(i.toString())
                    val schedule_sp = getSharedPreferences("schedule", Context.MODE_PRIVATE)
                    val schedule_str = schedule_sp.getString(getFullDate, "")
                    if (schedule_str != getString(R.string.no_schedule)) {
                        nextSchedule = schedule_str
                        nextScheduleDay = getDate
                        break
                    }
                }
                runOnUiThread {
                    if (nextSchedule == null || nextSchedule.isEmpty()) {
                        item_schedule.text = getString(R.string.no_next_schedule)
                    } else {
                        item_schedule.text = String.format(
                            getString(R.string.next_schedule),
                            nextSchedule,
                            nextScheduleDay
                        )
                    }
                }
            }.start()
            loadingProgress.visibility = View.GONE
        }


    override fun onResume() {
        super.onResume()
        nav_view.setCheckedItem(R.id.nav_home)
        Log.d("GGCJ", "onResume")
        val sp = getSharedPreferences("AppIntro", Context.MODE_PRIVATE)
        if (sp.getBoolean("first", false)) {
            Log.d("GGCJ", "setView")
            Log.d("GGCJ", sp.getBoolean("first", false).toString())
            setView()
        }
    }

    override fun onDestroy() {
        super.onDestroy()
        unregisterReceiver(mBroadcastReceiver)
    }
}